lib/types.nix

```diff
@@ -770,6 +772,21 @@ rec {
       nestedTypes.elemType = elemType;
     };
 
+    # Network types
+    ipv6AddrStr = elemType: mkOptionType {
+      name = "Valid IPv6 address string";
+      descriptionClass = "noun";
+      check = ipv6.isValidIpStr;
+      merge = mergeEqualOption;
+    };
+
+    ipv6Addr = elemType: mkOptionType {
+      name = "Attribute set of parsed IPv6 address";
+      descriptionClass = "noun";
+      check = v: isAttrs v && all (el: hasAttr el v) ["address" "addressCidr" "prefixLength" "_address" ];
+      merge = mergeEqualOption;
+    };
```

Comment:
I think it's fine if we only have the *Str types and recommend users of the types to also do an `apply`, in practice this would look something like:
```nix
listenAddress = lib.mkOption {
  type = lib.types.ipv6AddrStr;
  apply = fromString;
  ...
};
```

And then whenever `config.service.foo.listenAddress` gets accessed, it returns the `attrset`
We can also set the `outPath` key in the attrSet returned by `fromString` to the initial input, means that whenever the user tries to just do a string substitution like `"${config.service.foo.listenAddress}"` they get the same value they put in, but they still can just access and treat the value like a normal attrset.

We could also write a function like:
```nix
let
  lib.mkNetworkingOption = attrSet: lib.mkOption (attrSet // {
    type = lib.types.ipv${attrSet.apply}AddrStr;
    apply = x: attrSet.apply (lib.fromString x);
  });
```
and tell people they are supposed to use those and keep the types internal.
*(this is just a rough sketch of what this could look like, there are quite a few more checks that should be done here to e.g. give proper error messages)

cc @roberth here since I'm not quite sure what the best solution for this problem would be and believe you have more experience with the module system.

Reply:
I'd generally recommend `types.coercedTo` over `apply`.
They achieve the same effect, but a type creates a more consistent experience between modules, and is easier to use than a separately specified functions, particularly when nesting into other types, such as `oneOf`.

> `outPath`

Not sure; only if it is ok to always present it the same way, which I believe might not be ideal for IPv6, which looks different when used in URLs.
Whereas `"http://${ip}"` was safe in IPv4, we should use something like `"http://${toURLHost ip}"` for IPv6 and mixed use cases. However, we don't want that behavior for e.g. some network routing configuration, where the brackets are not expected.
By omitting `outPath`, we force module authors to look up the address rendering functions, and make the right decision for their use case.

Reply:
Another small benefit of a type is that it can behave like `coercedTo` and thereby avoid repeatedly parsing and rendering when the option value is forwarded to other options.

Reply:
AFAIK ipv6 looks the same in urls, the only difference is having to put ports in brackets, which you also do in other environments. Do you have an example of the different v6 representations?